diff --git a/arch/arm/dts/stm32f7-pinctrl.dtsi b/arch/arm/dts/stm32f7-pinctrl.dtsi
index fe4cfda72a..9fae1ea5e0 100644
--- a/arch/arm/dts/stm32f7-pinctrl.dtsi
+++ b/arch/arm/dts/stm32f7-pinctrl.dtsi
@@ -157,7 +157,7 @@
 					slew-rate = <0>;
 				};
 				pins2 {
-					pinmux = <STM32_PINMUX('B', 7, AF7)>; /* USART1_RX */
+					pinmux = <STM32_PINMUX('A', 10, AF7)>; /* USART1_RX */
 					bias-disable;
 				};
 			};
diff --git a/arch/arm/dts/stm32f769-disco-u-boot.dtsi b/arch/arm/dts/stm32f769-disco-u-boot.dtsi
index 7dfe430a40..adbce24c81 100644
--- a/arch/arm/dts/stm32f769-disco-u-boot.dtsi
+++ b/arch/arm/dts/stm32f769-disco-u-boot.dtsi
@@ -100,9 +100,9 @@
 	/* Memory configuration from sdram datasheet MT48LC_4M32_B2B5-6A */
 	bank1: bank@0 {
 		u-boot,dm-pre-reloc;
-		st,sdram-control = /bits/ 8 <NO_COL_8
-					     NO_ROW_12
-					     MWIDTH_32
+		st,sdram-control = /bits/ 8 <NO_COL_9
+					     NO_ROW_13
+					     MWIDTH_16
 					     BANKS_4
 					     CAS_3
 					     SDCLK_2
@@ -138,24 +138,7 @@
 
 	fmc_pins: fmc@0 {
 		pins {
-			pinmux = <STM32_PINMUX('I',10, AF12)>, /* D31 */
-				 <STM32_PINMUX('I', 9, AF12)>, /* D30 */
-				 <STM32_PINMUX('I', 7, AF12)>, /* D29 */
-				 <STM32_PINMUX('I', 6, AF12)>, /* D28 */
-				 <STM32_PINMUX('I', 3, AF12)>, /* D27 */
-				 <STM32_PINMUX('I', 2, AF12)>, /* D26 */
-				 <STM32_PINMUX('I', 1, AF12)>, /* D25 */
-				 <STM32_PINMUX('I', 0, AF12)>, /* D24 */
-				 <STM32_PINMUX('H',15, AF12)>, /* D23 */
-				 <STM32_PINMUX('H',14, AF12)>, /* D22 */
-				 <STM32_PINMUX('H',13, AF12)>, /* D21 */
-				 <STM32_PINMUX('H',12, AF12)>, /* D20 */
-				 <STM32_PINMUX('H',11, AF12)>, /* D19 */
-				 <STM32_PINMUX('H',10, AF12)>, /* D18 */
-				 <STM32_PINMUX('H', 9, AF12)>, /* D17 */
-				 <STM32_PINMUX('H', 8, AF12)>, /* D16 */
-
-				 <STM32_PINMUX('D',10, AF12)>, /* D15 */
+			pinmux = <STM32_PINMUX('D',10, AF12)>, /* D15 */
 				 <STM32_PINMUX('D', 9, AF12)>, /* D14 */
 				 <STM32_PINMUX('D', 8, AF12)>, /* D13 */
 				 <STM32_PINMUX('E',15, AF12)>, /* D12 */
@@ -172,14 +155,13 @@
 				 <STM32_PINMUX('D',15, AF12)>, /* D1 */
 				 <STM32_PINMUX('D',14, AF12)>, /* D0 */
 
-				 <STM32_PINMUX('I', 5, AF12)>, /* NBL3 */
-				 <STM32_PINMUX('I', 4, AF12)>, /* NBL2 */
 				 <STM32_PINMUX('E', 1, AF12)>, /* NBL1 */
 				 <STM32_PINMUX('E', 0, AF12)>, /* NBL0 */
 
 				 <STM32_PINMUX('G', 5, AF12)>, /* BA1 */
 				 <STM32_PINMUX('G', 4, AF12)>, /* BA0 */
 
+				 <STM32_PINMUX('G', 2, AF12)>, /* A12 */
 				 <STM32_PINMUX('G', 1, AF12)>, /* A11 */
 				 <STM32_PINMUX('G', 0, AF12)>, /* A10 */
 				 <STM32_PINMUX('F',15, AF12)>, /* A9 */
@@ -193,11 +175,11 @@
 				 <STM32_PINMUX('F', 1, AF12)>, /* A1 */
 				 <STM32_PINMUX('F', 0, AF12)>, /* A0 */
 
-				 <STM32_PINMUX('H', 3, AF12)>, /* SDNE0 */
-				 <STM32_PINMUX('H', 5, AF12)>, /* SDNWE */
+				 <STM32_PINMUX('C', 2, AF12)>, /* SDNE0 */
+				 <STM32_PINMUX('C', 0, AF12)>, /* SDNWE */
 				 <STM32_PINMUX('F',11, AF12)>, /* SDNRAS */
 				 <STM32_PINMUX('G',15, AF12)>, /* SDNCAS */
-				 <STM32_PINMUX('H', 2, AF12)>, /* SDCKE0 */
+				 <STM32_PINMUX('C', 3, AF12)>, /* SDCKE0 */
 				 <STM32_PINMUX('G', 8, AF12)>; /* SDCLK> */
 			slew-rate = <2>;
 		};
@@ -228,7 +210,7 @@
 
 &qspi {
 	reg = <0xA0001000 0x1000>, <0x90000000 0x4000000>;
-	flash0: mx66l51235l@0 {
+	flash0: w25q256@0 {
 		#address-cells = <1>;
 		#size-cells = <1>;
 		compatible = "jedec,spi-nor";
diff --git a/arch/arm/dts/stm32f769-disco.dts b/arch/arm/dts/stm32f769-disco.dts
index 03cfbd7cc2..ed5ad2fe72 100644
--- a/arch/arm/dts/stm32f769-disco.dts
+++ b/arch/arm/dts/stm32f769-disco.dts
@@ -11,7 +11,7 @@
 #include <dt-bindings/gpio/gpio.h>
 
 / {
-	model = "STMicroelectronics STM32F769-DISCO board";
+	model = "STMicroelectronics STM32F767-APOLLO board";
 	compatible = "st,stm32f769-disco", "st,stm32f769";
 
 	chosen {
@@ -21,7 +21,7 @@
 
 	memory@c0000000 {
 		device_type = "memory";
-		reg = <0xC0000000 0x1000000>;
+		reg = <0xC0000000 0x2000000>;
 	};
 
 	aliases {
@@ -31,7 +31,7 @@
 	leds {
 		compatible = "gpio-leds";
 		led-green {
-			gpios = <&gpioj 5 GPIO_ACTIVE_HIGH>;
+			gpios = <&gpiob 1 GPIO_ACTIVE_HIGH>;
 			linux,default-trigger = "heartbeat";
 		};
 		led-red {
