diff --git a/arch/arm/boot/dts/stm32f746-disco.dts b/arch/arm/boot/dts/stm32f746-disco.dts
index 4d85dba59e1d..066bf396a8a5 100644
--- a/arch/arm/boot/dts/stm32f746-disco.dts
+++ b/arch/arm/boot/dts/stm32f746-disco.dts
@@ -45,7 +45,7 @@
 #include <dt-bindings/input/input.h>
 
 / {
-	model = "STMicroelectronics STM32F746-DISCO board";
+	model = "STMicroelectronics STM32F767IGT6-LQFP176 board";
 	compatible = "st,stm32f746-disco", "st,stm32f746";
 
 	chosen {
@@ -54,7 +54,7 @@
 	};
 
 	memory {
-		reg = <0xC0000000 0x800000>;
+		reg = <0xD0000000 0x800000>;
 	};
 
 	aliases {
@@ -82,7 +82,7 @@
 };
 
 &usart1 {
-	pinctrl-0 = <&usart1_pins_b>;
+	pinctrl-0 = <&usart1_pins_a>;
 	pinctrl-names = "default";
 	status = "okay";
 };
@@ -91,7 +91,7 @@
 	dr_mode = "host";
 	pinctrl-0 = <&usbotg_fs_pins_a>;
 	pinctrl-names = "default";
-	status = "okay";
+	status = "disabled";
 };
 
 &usbotg_hs {
@@ -100,5 +100,5 @@
 	phy-names = "usb2-phy";
 	pinctrl-0 = <&usbotg_hs_pins_b>;
 	pinctrl-names = "default";
-	status = "okay";
+	status = "disabled";
 };
