diff --git a/arch/arm/boot/dts/stm32f7-pinctrl.dtsi b/arch/arm/boot/dts/stm32f7-pinctrl.dtsi
index fe4cfda72a47..8a9d77862715 100644
--- a/arch/arm/boot/dts/stm32f7-pinctrl.dtsi
+++ b/arch/arm/boot/dts/stm32f7-pinctrl.dtsi
@@ -284,6 +284,18 @@ pins2 {
 					slew-rate = <2>;
 				};
 			};
+
+			qspi_pins: qspi@0 {
+				pins {
+					pinmux = <STM32_PINMUX('B', 2, AF9)>, /* CLK */
+						 <STM32_PINMUX('B', 6, AF10)>, /* BK1_NCS */
+						 <STM32_PINMUX('F', 8, AF10)>, /* BK1_IO0 */
+						 <STM32_PINMUX('F', 9, AF10)>, /* BK1_IO1 */
+						 <STM32_PINMUX('F', 6, AF9)>, /* BK1_IO3 */
+						 <STM32_PINMUX('F', 7, AF9)>; /* BK1_IO2 */
+					slew-rate = <2>;
+				};
+			};
 		};
 	};
 };
diff --git a/arch/arm/boot/dts/stm32f746.dtsi b/arch/arm/boot/dts/stm32f746.dtsi
index 93c063796780..a0ed5fbaf6f5 100644
--- a/arch/arm/boot/dts/stm32f746.dtsi
+++ b/arch/arm/boot/dts/stm32f746.dtsi
@@ -639,6 +639,21 @@ usbotg_fs: usb@50000000 {
 			clock-names = "otg";
 			status = "disabled";
 		};
+
+		qspi: spi@A0001000 {
+			compatible = "st,stm32f469-qspi";
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0xA0001000 0x1000>, <0x90000000 0x10000000>;
+			reg-names = "qspi", "qspi_mm";
+			interrupts = <91>;
+			spi-max-frequency = <108000000>;
+			clocks = <&rcc 0 STM32F7_AHB3_CLOCK(QSPI)>;
+			resets = <&rcc STM32F7_AHB3_RESET(QSPI)>;
+			pinctrl-0 = <&qspi_pins>;
+			pinctrl-names = "default";
+			status = "disabled";
+		};
 	};
 };
 
diff --git a/arch/arm/boot/dts/stm32f769-disco.dts b/arch/arm/boot/dts/stm32f769-disco.dts
index 1626e00bb2cb..6496599a8b23 100644
--- a/arch/arm/boot/dts/stm32f769-disco.dts
+++ b/arch/arm/boot/dts/stm32f769-disco.dts
@@ -47,7 +47,7 @@
 #include <dt-bindings/gpio/gpio.h>
 
 / {
-	model = "STMicroelectronics STM32F769-DISCO board";
+	model = "STMicroelectronics STM32F767-APOLLO board";
 	compatible = "st,stm32f769-disco", "st,stm32f769";
 
 	chosen {
@@ -57,7 +57,7 @@ chosen {
 
 	memory@c0000000 {
 		device_type = "memory";
-		reg = <0xC0000000 0x1000000>;
+		reg = <0xC0000000 0x2000000>;
 	};
 
 	aliases {
@@ -67,7 +67,7 @@ aliases {
 	leds {
 		compatible = "gpio-leds";
 		green {
-			gpios = <&gpioj 5 GPIO_ACTIVE_HIGH>;
+			gpios = <&gpiob 1 GPIO_ACTIVE_HIGH>;
 			linux,default-trigger = "heartbeat";
 		};
 		red {
@@ -153,3 +153,36 @@ &usbotg_hs {
 	pinctrl-names = "default";
 	status = "okay";
 };
+
+&qspi {
+	reg = <0xA0001000 0x1000>, <0x90000000 0x2000000>;
+	status = "okay";
+	qflash0: w25q256@0 {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		compatible = "winboud,w25q256", "jedec,spi-nor";
+		spi-max-frequency = <108000000>;
+		spi-rx-bus-width = <4>;
+		reg = <0x0>;
+
+		partition@20000 {
+			label = "device-tree";
+			reg = <0x0 0x20000>;
+		};
+
+		partition@400000 {
+			label = "linux";
+			reg = <0x20000 0x400000>;
+		};
+
+		partition@420000 {
+			label = "rootfs";
+			reg = <0x420000 0x400000>;
+		};
+
+		partition@820000 {
+			label = "jffs2";
+			reg = <0x820000 0x1000000>;
+		};
+	};
+};
\ No newline at end of file
